### ISSUE TITLE
Is it possible to obtain batch size in keras layer

### ISSUE THREAD

**Issue by siddheshk** [fchollet/keras#5211]
I need to perform a different operation on each tensor slice. Therefore I need to know the batch size inside a keras layer. The call function of my layer is written below:

```python
    def call(self, inputs, mask=None):
        x = inputs
        input_shape = K.int_shape(inputs[0])
        timesteps = input_shape[1]
        x = K.reshape(x, (-1,)+ input_shape[2:])
        new_input_shape = K.int_shape(x)
        for i in range(new_input_shape[0]): #cant do this currently
            p1 = x[i, :, :, :] #some slice operation
           
```

Currently the code does not go into the loop as new_input_shape[0] is None. Can I somehow get the batch size?

Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [ ] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by bstriner** [comment]
`Call` is only called once or twice. It makes a symbolic graph that is run later many times. So the shape is None when it is originally called but will be an actual value when used later.

Use K.shape instead of K.int_shape and it should work.

**Comment by bstriner** [comment]
K.int_shape checks keras metadata for the shape but K.shape gets the actual shape from the TF/TH tensor.

* K.shape is not helpful for debugging because it is just a symbol, but it is the one to actually use in your code. 
* K.int_shape is helpful for debugging and checking, but is just some static metadata and won't normally reflect the current batch size.

Cheers

**Comment by siddheshk** [comment]
@bstriner Thanks for your help. I tried using K.shape instead of K.int_shape. But `range` still throws a TypeError saying that an integer is required.

**Comment by bstriner** [comment]
Ah! Didn't see that was where you were having your issue. Read up on `scan`, available in both theano and tensorflow, but no scan in Keras that works on both. What backend are you on?

Scan applies an arbitrary function along the first axis. So all you need to do is write a function for your element-wise operation and call scan. No for loop required. Is that what you need?

**Comment by farizrahman4u** [comment]
You can use K.rnn, which will work on both backends.

**Comment by bstriner** [comment]
@farizrahman4u good idea. It would be nice to have something a little more generic than K.rnn but that is an option. I think it has some checks for 3d data, and applies the scan to the 2nd dimension, so he might have to reformat his data slightly.

**Comment by siddheshk** [comment]
I am on theano backend.
The loop is currently needed to index another input which will provide the indexes for slicing. Something like 
```python
for i in xrange(shape[0]):
    p1 = x[i, :, :entity_location[0, i, 0], :]
```
Here entity_location is another TensorVariable that will give me the index.
Even if I use a RNN, I would somehow need the variable `i` to index the second TensorVariable. Am I missing something?

**Comment by farizrahman4u** [comment]
You can use a state to track the index and increment it every step.

**Comment by siddheshk** [comment]
Using states might just work. Thanks for your help. But this seems to be a lot of effort just to work around the fact that I cannot obtain the batch size directly. Is it possible to have a better way of doing this in the future?

**Comment by bstriner** [comment]
Another possibility is to arrange both your input and your slicing array so
the batch dim is the first dim. You can pass both to scan. You shouldn't
need the batch dim that way.

**Comment by bstriner** [comment]
Something like this. Make sure the shape of each iteration of output coming out of your scan is the same.

```python
entity_location = K.permute_dimensions((1,0,2))
y, _ = theano.scan(lambda _x, _el: _x[:, _el[0, 0], :], sequences = (x, entity_location))
```

**Comment by siddheshk** [comment]
@bstriner Thanks. This makes it a lot easier :)
I will close this thread for now.

**Comment by Manezx** [comment]
Is it possible to make the same with tf.scan ?
it seems like 'elems' take only one Tensor as input.
